core/src/main/scala/filodb.core/memstore/ratelimit/RocksDbCardinalityStore.scala

```diff
@@ -144,108 +160,129 @@ class RocksDbCardinalityStore(ref: DatasetRef, shard: Int) extends CardinalitySt
+    }
+    if (keyDepth > shardKeyPrefix.size) {
+      b.append(KeySeparator)
     }
+    b.toString()
   }
 
-  private def cardinalityToBytes(card: Cardinality): Array[Byte] = {
+  /**
+   * Builds keys as:
+   *   <keyDepth>{KeySeparator}<name-1>{KeySeparator}<name-2> ... <name-n>
+   * The final {KeySeparator} is always omitted.
+   */
+  private def toStringKey(shardKeyPrefix: Seq[String]): String = {
+    toStringKeyPrefix(shardKeyPrefix, shardKeyPrefix.size)
+  }
+
+  private def cardinalityNodeToBytes(card: CardinalityNode): Array[Byte] = {
     val out = new Output(500)
     kryo.get().writeObject(out, card)
     out.close()
     out.toBytes
   }
 
-  private def bytesToCardinality(bytes: Array[Byte]): Cardinality = {
+  private def bytesToCardinalityNode(bytes: Array[Byte]): CardinalityNode = {
     val inp = new Input(bytes)
-    val c = kryo.get().readObject(inp, classOf[Cardinality])
+    val c = kryo.get().readObject(inp, classOf[CardinalityNode])
     inp.close()
     c
   }
 
-  override def store(shardKeyPrefix: Seq[String], card: Cardinality): Unit = {
-    val key = toStringKey(shardKeyPrefix, false).getBytes(StandardCharsets.UTF_8)
+  override def store(card: Cardinality): Unit = {
+    val key = toStringKey(card.prefix).getBytes(StandardCharsets.UTF_8)
     logger.debug(s"Storing shard=$shard dataset=$ref ${new String(key)} with $card")
-    db.put(key, cardinalityToBytes(card))
+    db.put(key, cardinalityNodeToBytes(CardinalityNode.fromCardinality(card)))
   }
 
   def getOrZero(shardKeyPrefix: Seq[String], zero: Cardinality): Cardinality = {
-    val value = db.get(toStringKey(shardKeyPrefix, false).getBytes(StandardCharsets.UTF_8))
-    if (value == NotFound) zero else bytesToCardinality(value)
+    val value = db.get(toStringKey(shardKeyPrefix).getBytes(StandardCharsets.UTF_8))
+    if (value == NotFound) zero else CardinalityNode.toCardinality(bytesToCardinalityNode(value), shardKeyPrefix)
   }
 
   override def remove(shardKeyPrefix: Seq[String]): Unit = {
-    db.delete(toStringKey(shardKeyPrefix, false).getBytes(StandardCharsets.UTF_8))
+    db.delete(toStringKey(shardKeyPrefix).getBytes(StandardCharsets.UTF_8))
   }
 
-  override def scanChildren(shardKeyPrefix: Seq[String]): Seq[Cardinality] = {
-    val it = db.newIterator()
-    val buf = ArrayBuffer[Cardinality]()
-    try {
-      val searchPrefix = toStringKey(shardKeyPrefix, true)
-      logger.debug(s"Scanning shard=$shard dataset=$ref ${new String(searchPrefix)}")
-      it.seek(searchPrefix.getBytes(StandardCharsets.UTF_8))
-      import scala.util.control.Breaks._
-
-      breakable {
-        while (it.isValid()) {
-          val key = new String(it.key(), StandardCharsets.UTF_8)
-          if (key.startsWith(searchPrefix)) {
-            buf += bytesToCardinality(it.value())
-          } else break // dont continue beyond valid results
-          it.next()
+  override def scanChildren(shardKeyPrefix: Seq[String], depth: Int): CardIter = {
```

Comment:
Need to document in BOLD that the iterator needs to be closed after consumption. Else there will be an off-heap leak that is extremely hard to identify. Were you able to verify if it will be closed?

IIRC, this was one reason I did not choose iterators for this. But if we can be assured that we will safely close then I am ok with this.

Reply:
After looking at the details, the benefit of the iterator would be reduced if we constructed `CardinalityRecord` here per my other suggestion. We can simplify and return a Seq here instead of the iterator close complexity.

Reply:
Done!